src/server/game/Skills/SkillExtraItems.cpp

```diff
@@ -20,11 +20,97 @@
 #include "DatabaseEnv.h"
 #include "Log.h"
 #include "Player.h"
+#include "ObjectMgr.h"
 #include <map>
 
 // some type definitions
 // no use putting them in the header file, they're only used in this .cpp
 
+// struct to store information about perfection procs
+// one entry per spell
+struct SkillPerfectItemEntry
+{
+    // the spell id of the spell required - it's named "specialization" to conform with SkillExtraItemEntry
+    uint32 requiredSpecialization;
+    // perfection proc chance
+    float perfectCreateChance;
+    // itemid of the resulting perfect item
+    uint32 perfectItemType;
+
+    SkillPerfectItemEntry()
+        : requiredSpecialization(0), perfectCreateChance(0.0f), perfectItemType(0) { }
+    SkillPerfectItemEntry(uint32 rS, float pCC, uint32 pIT)
+        : requiredSpecialization(rS), perfectCreateChance(pCC), perfectItemType(pIT) { }
+};
+
+// map to store perfection info. key = spellId of the creation spell, value is the perfectitementry as specified above
+typedef std::map<uint32, SkillPerfectItemEntry> SkillPerfectItemMap;
+
+SkillPerfectItemMap SkillPerfectItemStore;
+
+// loads the perfection proc info from DB
+void LoadSkillPerfectItemTable()
+{
+    uint32 oldMSTime = getMSTime();
+
+    SkillPerfectItemStore.clear(); // reload capability
+
+    //                                                  0               1                      2                  3
+    QueryResult result = WorldDatabase.Query("SELECT spellId, requiredSpecialization, perfectCreateChance, perfectItemType FROM skill_perfect_item_template");
+
+    if (!result) {
```

Comment:
TC codestyle suggests that you move the opening bracket { to a new line and align it with its closing bracket } .